lib/agent0/agent0/hyperdrive/exec/execute_agent_trades.py

```diff
@@ -39,28 +45,69 @@ async def async_execute_single_agent_trade(
         The HyperdriveAgent that is conducting the trade
     hyperdrive : HyperdriveInterface
         The Hyperdrive API interface object
+
+    Returns
+    -------
     """
     trades: list[types.Trade[HyperdriveMarketAction]] = agent.get_trades(interface=hyperdrive)
-    for trade_object in trades:
-        policy_name = agent.policy.__class__.__name__
-        logging.info(
-            "AGENT %s (%s) to perform %s for %g",
-            str(agent.checksum_address),
-            policy_name,
-            trade_object.market_action.action_type,
-            float(trade_object.market_action.trade_amount),
-        )
-        try:
-            wallet_deltas = await async_match_contract_call_to_trade(agent, hyperdrive, trade_object)
-            agent.wallet.update(wallet_deltas)
-        except UnknownBlockError as exc:
-            logging.error(exc)
+
+    # Make trades async for this agent. This way, an agent can submit multiple trades for a single block
+    # This unblocks any other agents running here as well, as previously, an agent would take one block per trade
+    # which blocks all other agents from making trades until all trades are finished.
+
+    # TODO preliminary search shows async tasks has very low overhead:
+    # https://stackoverflow.com/questions/55761652/what-is-the-overhead-of-an-asyncio-task
+    # However, should probably test what the limit number of trades an agent can make in one block
+    wallet_deltas_or_exception: list[HyperdriveWalletDeltas | Exception] = await asyncio.gather(
+        *[async_match_contract_call_to_trade(agent, hyperdrive, trade_object) for trade_object in trades],
+        # Instead of throwing exception, return the exception to the caller here
+        return_exceptions=True,
+    )
+
+    # TODO Here, gather returns results based on original order of trades, but this order isn't guaranteed
+    # because of async. Ideally, we should return results based on the order of trades. Can we use nonce here
+    # to see order?
+
+    # Sanity check
+    assert len(wallet_deltas_or_exception) == len(trades)
```

Comment:
I think it's better to say 
```
if len(wallet_deltas_or_exception) != len(trades):
    raise AssertionError(f"The number of wallet deltas should match the number of trades, but does not.\n{wallet_deltas_or_exception=}\n{trades=}")
 ```
 
 I think these conditional checks are generally faster than assertions, and it's nice to get a more detailed error message.